pandas/core/arrays/datetimes.py

```diff
@@ -828,7 +818,34 @@ def _add_offset(self, offset: BaseOffset) -> Self:
             result = type(self)._from_sequence(res_values, dtype=self.dtype)
 
         else:
+            units = [
+                "ns",
+                "us",
+                "ms",
+                "s",
+            ]
+            res_unit = self.unit
+            if type(offset) is DateOffset:
+                nano = offset.kwds.get("nanoseconds", 0)
+                micro = offset.kwds.get("microseconds", 0)
+                if nano:
+                    res_unit = "ns"
+                elif micro and self.unit != "ns":
+                    res_unit = "us"
```

Comment:
side-note for another PR: might be worth defining a `unit` property on DateOffset since i think this recently came up in date_range too.